# Introduction to Pandas Library
- Pandas is an open source library in python which is know for its rich applications and utilities for all kinds of mathematical, financial and statistical functions
- It is useful in data manipulation and analysis
- It provides fast, flexible, and expressive data structures designed to make working with structured (tabular, multidimensional, potentially heterogeneous) and time series data



#### Installing pandas

In [ ]:
!pip install pandas

#### Importing pandas

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Series

#### Series -

   - are one-dimensional ndarray with axis labels (homogenous data)
   - labels need not be unique but must be of immutable type



### Creating Series

###### Ex. Create series using the given list of names

In [ ]:
names = pd.Series(["Jack", "Jane", "George"])
names

In [ ]:
names.index  # labels of the series

In [ ]:
names.values # Values in the series

##### Note - An ordered sequence eg - list, tuple, dict, array can only be converted into a series

In [ ]:
salaries = pd.Series(np.random.randint(30000, 60000, 5))
salaries

###### Assign names as labels

In [ ]:
salaries.index = ["Jane", "Jack", "George", "Rosie", "Dori"]
salaries

In [ ]:
# Assigning labels while creating series
salaries = pd.Series(np.random.randint(30000, 60000, 5), index = ["Jane", "Jack", "George", "Rosie", "Dori"])

### Extracting elements from series

#### Indexing

In [ ]:
salaries.iloc[1]  # indexing based on index position

In [ ]:
salaries["Jane"]  # extracting value based on labels

#### Slicing

In [ ]:
salaries.iloc[0:3]

In [ ]:
salaries["Jane" : "George"]

#### Conditional Indexing

In [ ]:
salaries[salaries > 50000]

### Operations on Series

###### Ex. Increment the salaries by 10%

In [ ]:
salaries * 1.10

### Ranking and Sorting

- series.sort_values(`ascending=True`, `inplace=False`, `na_position = {"first","last"}`)
- series.sort_index(`ascending=True`, `inplace=False`)
- series.rank(`ascending=False`, `method={"average","min","dense"}`, `na_option = {"top","bottom"}`)

In [ ]:
salaries = pd.Series(np.random.randint(30000, 60000, 5), index = ["Jane", "Jack", "George", "Rosie", "Dori"])
salaries = pd.concat((salaries, pd.Series([np.nan, np.nan], index = ["Janet", "Sam"])))
salaries

###### Ex. Sort by values

In [ ]:
salaries.sort_values(ascending=False, na_position="first", ignore_index=True)

###### Ex. Sort by index

In [ ]:
salaries.sort_index(ascending=False)

In [ ]:
salaries

###### Ex. Rank the series

In [ ]:
salaries.rank(method="min", na_option="bottom", ascending=False).astype(int)

In [ ]:
salaries

In [ ]:
marks = pd.Series([80, 90, 80, 70, 60, 60, 50])
marks

In [ ]:
marks.rank(ascending=False, method="min").astype(int)

##### Note - to modify original series/dataframe inplace can be set to True

### Working with NULLs

In [ ]:
salaries.isna()

In [ ]:
salaries.isna().sum()

In [ ]:
salaries.isna().any()

In [ ]:
salaries.isna().all()

In [ ]:
salaries.fillna(0)

In [ ]:
salaries.ffill()

In [ ]:
salaries.bfill()

<hr><hr>

# Dataframe

A DataFrame is two dimensional data structure where the data is arranged in the tabular format in rows and columns

#### DataFrame features:

- Columns can be of different data types
- Size of dataframe can be changes
- Axes(rows and columns) are labeled
- Arithmetic operations can be performed on rows and columns

### Creating Dataframes

In [ ]:
employees = {"Name" : ["Jack", "Bill", "Lizie", "Jane", "George"],
            "Designation" : ["HR", "Manager", "Developer", "Intern", "Manager"],
            "Salary": [40000, 60000, 25000, 12000, 70000]}

df = pd.DataFrame(employees)
df

### Accessing Dataframes

In [ ]:
df["Name"]

In [ ]:
df.Name

### Operations on dataframes

###### Ex. Average Salary

In [ ]:
df.Salary.mean()

###### Ex. Average Salary of managers

In [ ]:
df[df.Designation == "Manager"]

### Concataneting and Merging Dataframes

In [ ]:
df_jan = pd.DataFrame({"Order ID" : range(101, 111), "Sales" : np.random.randint(10000, 50000, 10)})
df_feb = pd.DataFrame({"Order ID" : range(111, 121), "Sales" : np.random.randint(10000, 50000, 10)})
df_mar = pd.DataFrame({"Order ID" : range(121, 131), "Sales" : np.random.randint(10000, 50000, 10)})

#### Concatenate
pd.concat(`tuple of dfs`, `ignore_index = False`, `axis=0`)

In [ ]:
pd.concat((df_jan, df_feb, df_mar), ignore_index=True)

In [ ]:
pd.concat((df_jan, df_feb, df_mar), axis=1)

#### Merging Dataframes

`df1.merge(df2, how="", left_on="", right_on="", left_index= "" , right_index="")`

In [ ]:
df_emp = pd.DataFrame({"Name" : ["Jack", "Bill", "Lizie", "Jane", "George"],
            "Designation" : ["HR", "Manager", "Developer", "Intern", "Manager"]})
df_emp

In [ ]:
base_salaries = pd.DataFrame({"Post" : ["HR", "Developer", "Manager", "Senior Manager"],
            "Salary": [40000, 25000, 70000, 1000000]})
base_salaries

#### Inner Merge - returns rows present in both tables

In [ ]:
df_emp.merge(base_salaries, how="inner", left_on="Designation", right_on="Post") # use on = "Designation" if both dfs have same colunm name as Designation

#### Left Merge - returns data from left table and corresponding data from right, returns NAN for non matching values

In [ ]:
df_emp.merge(base_salaries, how="left", left_on="Designation", right_on="Post") # use on = "Designation" if both dfs have same colunm name as Designation

#### Right Merge

In [ ]:
df_emp.merge(base_salaries, how="right", left_on="Designation", right_on="Post") # use on = "Designation" if both dfs have same colunm name as Designation

#### Outer Merge

In [ ]:
df_emp.merge(base_salaries, how="outer", left_on="Designation", right_on="Post") # use on = "Designation" if both dfs have same colunm name as Designation

## Reading data from Data Sources

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Reading data from MYSQL or SQLITE3

In [ ]:
pip install sqlalchemy

###### Syntax for MSSQL

connection_string = f"mssql+pyodbc://{username}:{password}@{server}/{database}"


In [37]:
from sqlalchemy import create_engine
conn = create_engine(r"sqlite:///employee.sqlite3")

In [34]:
conn

Engine(sqlite:///employee.sqlite3)

In [ ]:
pd.read_sql("Employee", conn)

In [38]:
df = pd.read_sql_query("Select * from Employee where Designation = 'Manager'", conn)
df

index    Name  Salary Designation  Age
0      0  Claire   88962     Manager   35
1      2    Sean  117501     Manager   36
2     11  Sandra  115116     Manager   41
3     14   Tracy  109132     Manager   34
4     15    Matt   83327     Manager   43

In [ ]:
df.to_sql("Employee", conn, if_exists="replace")

### Examples using Coffee Shop Dataset

###### Ex. Connecting to Excel File

In [ ]:
pd.read_excel("filename.xlsx", sheet_name="sheet_name")  # demo syntax

##### Note - may generate ModuleNotFoundError - openpyxl. 
pip install openpyxl

###### Ex. Read data from `coffee_sales.csv`

In [39]:
df = pd.read_csv("coffee_sales.csv")
df

This is dummy data Unnamed: 1 Unnamed: 2 Unnamed: 3        Unnamed: 4  \
0                    NaN        NaN        NaN        NaN               NaN   
1                    NaN        NaN        NaN        NaN               NaN   
2                    NaN       Date  Franchise       City           Product   
3                    NaN   1-Jan-21         M1     Mumbai          Amaretto   
4                    NaN   1-Feb-21         M1     Mumbai          Amaretto   
...                  ...        ...        ...        ...               ...   
4246                 NaN   1-Aug-26         M1     Mumbai  Regular Espresso   
4247                 NaN   1-Sep-26         M1     Mumbai  Regular Espresso   
4248                 NaN   1-Oct-26         M1     Mumbai  Regular Espresso   
4249                 NaN   1-Nov-26         M1     Mumbai  Regular Espresso   
4250                 NaN   1-Dec-26         M1     Mumbai  Regular Espresso   

     Unnamed: 5 Unnamed: 6     Unnamed: 7    Unnamed: 8  
0           NaN        NaN            NaN           NaN  
1           NaN        NaN            NaN           NaN  
2         Sales     Profit  Target Profit  Target Sales  
3       $219.00     $94.00        $100.00       $220.00  
4       $140.00     $34.00         $50.00       $140.00  
...         ...        ...            ...           ...  
4246    $719.00    $565.00        $390.00       $660.00  
4247     $81.00     -$4.00         $10.00        $70.00  
4248    $326.00    $129.00         $80.00       $290.00  
4249    $700.00    $463.00        $320.00       $640.00  
4250     $48.00    -$24.00         $10.00        $40.00  

[4251 rows x 9 columns]

### Cleaning DataFrame

###### Approach 1

pd.read_csv("coffee_sales.csv", header=3, usecols=[1, 5, 6], skiprows=range(4248, 4253))

In [58]:
df = pd.read_csv("coffee_sales.csv", header=3)  # Read Data
df

In [59]:
df.columns # get column names

Index(['Unnamed: 0', 'Date', 'Franchise', 'City', 'Product', 'Sales', 'Profit',
       'Target Profit', 'Target Sales'],
      dtype='object')

In [60]:
df.drop(columns= ['Unnamed: 0'], inplace=True)  # Drop column by name
df

Date Franchise    City           Product    Sales   Profit  \
0     1-Jan-21        M1  Mumbai          Amaretto  $219.00   $94.00   
1     1-Feb-21        M1  Mumbai          Amaretto  $140.00   $34.00   
2     1-Mar-21        M1  Mumbai          Amaretto  $145.00   -$2.00   
3     1-Apr-21        M1  Mumbai          Amaretto   $45.00   $11.00   
4     1-May-21        M1  Mumbai          Amaretto  $120.00   $13.00   
...        ...       ...     ...               ...      ...      ...   
4243  1-Aug-26        M1  Mumbai  Regular Espresso  $719.00  $565.00   
4244  1-Sep-26        M1  Mumbai  Regular Espresso   $81.00   -$4.00   
4245  1-Oct-26        M1  Mumbai  Regular Espresso  $326.00  $129.00   
4246  1-Nov-26        M1  Mumbai  Regular Espresso  $700.00  $463.00   
4247  1-Dec-26        M1  Mumbai  Regular Espresso   $48.00  -$24.00   

     Target Profit Target Sales  
0          $100.00      $220.00  
1           $50.00      $140.00  
2           $30.00      $180.00  
3           $20.00       $40.00  
4           $30.00      $120.00  
...            ...          ...  
4243       $390.00      $660.00  
4244        $10.00       $70.00  
4245        $80.00      $290.00  
4246       $320.00      $640.00  
4247        $10.00       $40.00  

[4248 rows x 8 columns]

###### Approach 2

In [76]:
df = pd.read_csv("coffee_sales.csv")  # Read Data 
df

This is dummy data Unnamed: 1 Unnamed: 2 Unnamed: 3        Unnamed: 4  \
0                    NaN        NaN        NaN        NaN               NaN   
1                    NaN        NaN        NaN        NaN               NaN   
2                    NaN       Date  Franchise       City           Product   
3                    NaN   1-Jan-21         M1     Mumbai          Amaretto   
4                    NaN   1-Feb-21         M1     Mumbai          Amaretto   
...                  ...        ...        ...        ...               ...   
4246                 NaN   1-Aug-26         M1     Mumbai  Regular Espresso   
4247                 NaN   1-Sep-26         M1     Mumbai  Regular Espresso   
4248                 NaN   1-Oct-26         M1     Mumbai  Regular Espresso   
4249                 NaN   1-Nov-26         M1     Mumbai  Regular Espresso   
4250                 NaN   1-Dec-26         M1     Mumbai  Regular Espresso   

     Unnamed: 5 Unnamed: 6     Unnamed: 7    Unnamed: 8  
0           NaN        NaN            NaN           NaN  
1           NaN        NaN            NaN           NaN  
2         Sales     Profit  Target Profit  Target Sales  
3       $219.00     $94.00        $100.00       $220.00  
4       $140.00     $34.00         $50.00       $140.00  
...         ...        ...            ...           ...  
4246    $719.00    $565.00        $390.00       $660.00  
4247     $81.00     -$4.00         $10.00        $70.00  
4248    $326.00    $129.00         $80.00       $290.00  
4249    $700.00    $463.00        $320.00       $640.00  
4250     $48.00    -$24.00         $10.00        $40.00  

[4251 rows x 9 columns]

**df.dropna(`axis = 0`, `how = "any"`, `inplace = False`)**
- axis 0 for row or 1 for column
- how - {any or all}

In [77]:
# Remove null rows
df.dropna(how="all", axis=0, inplace=True) # Delete a row with all null value - remove any null row in entire table, but header is not set
df.columns = df.iloc[0]

In [78]:
# Remove null cols
df.dropna(axis=1, how="all", inplace=True)

In [ ]:
df = df.iloc[1:]  # using slicing to extract all rows except first 

In [82]:
df.reset_index(drop= True, inplace=True)

In [83]:
df

2         Date Franchise    City           Product    Sales   Profit  \
0     1-Jan-21        M1  Mumbai          Amaretto  $219.00   $94.00   
1     1-Feb-21        M1  Mumbai          Amaretto  $140.00   $34.00   
2     1-Mar-21        M1  Mumbai          Amaretto  $145.00   -$2.00   
3     1-Apr-21        M1  Mumbai          Amaretto   $45.00   $11.00   
4     1-May-21        M1  Mumbai          Amaretto  $120.00   $13.00   
...        ...       ...     ...               ...      ...      ...   
4243  1-Aug-26        M1  Mumbai  Regular Espresso  $719.00  $565.00   
4244  1-Sep-26        M1  Mumbai  Regular Espresso   $81.00   -$4.00   
4245  1-Oct-26        M1  Mumbai  Regular Espresso  $326.00  $129.00   
4246  1-Nov-26        M1  Mumbai  Regular Espresso  $700.00  $463.00   
4247  1-Dec-26        M1  Mumbai  Regular Espresso   $48.00  -$24.00   

2    Target Profit Target Sales  
0          $100.00      $220.00  
1           $50.00      $140.00  
2           $30.00      $180.00  
3           $20.00       $40.00  
4           $30.00      $120.00  
...            ...          ...  
4243       $390.00      $660.00  
4244        $10.00       $70.00  
4245        $80.00      $290.00  
4246       $320.00      $640.00  
4247        $10.00       $40.00  

[4248 rows x 8 columns]

### Drop a column or row from dataframe

### Renaming a Columns

<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>

###### Rename Columns (column 5 - 8 are not accessible)

In [ ]:
headers = ["ShopID", "Year/Month", "Product", "Product Type", "State", "Target Profit", "Target Sales", "Profit", "Sales"]


#### Rename Single Column

### Understanding Data in Dataframe

- `df.shape` - gives the size of the dataframe in the format (row_count x column_count)
- `df.dtypes` - returns a Series with the data type of each column
- `df.info()` - prints information about a DataFrame including the index dtype and columns, non-null values and memory usage
- `df.head()` - prints the first 5 rows of you dataset including column header and the content of each row
- `df.tail()` - prints the last 5 rows of you dataset including column header and the content of each row

In [ ]:
df_coffee.shape

In [ ]:
df_coffee.dtypes

In [ ]:
df_coffee.info()

In [ ]:
df_coffee.head()

In [ ]:
df_coffee.head(3)

In [ ]:
df_coffee.tail()

In [ ]:
df_coffee.tail(3)

### Cleaning data

`df.apply()`

###### Convertinf Franchise into str

###### Ex. Converting Sales and Profits columns to float types

### Working with **null** values

`df.isna()` - Detect missing values. Return a boolean same-sized object indicating if the values are NA.

`df.fillna(value=None, inplace=False)` - Fill NA/NaN values using the specified method.

###### Ex. Identify Sales made by 'Caffe Latte'

### Removing Duplicate Data

### Replacing values

df.replace(old_value, new_value, inplace=True)

### Adding a new Column by calculation

###### Ex. Create columns showing `Sales` and `Profit` targets achieved

###### Ex. Count the number times Targets are achieved

### Creating a bar chart to view Target Status

#### using matplpotlib

#### using pandas

#### using seaborn

### Setting and Resetting Index

#### Seting Index
`df.set_index(keys, drop=True, inplace=False,)`- Set the DataFrame index (row labels) using one or more existing columns or arrays (of the correct length). The index can replace the existing index or expand on it.

#### Resetting Index

`df.reset_index(level=None, drop=False, inplace=False,)` - Reset the index of the DataFrame, and use the default one instead. If the DataFrame has a MultiIndex, this method can remove one or more levels.

### Indexing and Slicing using loc and iloc

#### Using loc to retrive data

- loc is label-based
- specify the name of the rows and columns that we need to filter out

###### Ex. Extract data for franchise 203

###### Ex. Extract `City` column

###### Ex. Extract `Sales` column for `Franchise - 203`

###### Ex. Extract `Sales` and `Profit` column for `Franchise - 203, 504`

#### Using iloc to retrive data

- iloc is integer index-based
- specify rows and columns by their integer index.

###### Ex. Extract row at index 2

###### Ex. Extract rows at index position 2,3,4

###### Ex. Extract column at index 0

###### Ex. Extract column from index 0 to 2

###### Ex. Extra rows 0 to 2 and columns 0 to 2

### Working with dates

#### Insert a column in between
df.insert(`index`, `column_name`, `default_value`)

###### Create columns Year and Month - extract data using pd.DatetimeIndex

###### Extract data for 2018

###### Extract data for Jan - 2018

###### Extract data for Jan - 2018 and 2019

###### Extract data starting from April - 2019

###### Extract data from Jan-2019 to Apr-2019

### Ranking and Sorting Dataframes

###### Ex. Rank the products in descending order of `Sales`

###### Ex. Sort the data in ascending order of `Rank`

### Grouping Dataframes

##### `df.groupby(by=None, as_index=True, sort=True, dropna=True)`

###### Ex. Find product wise total Sales - bar chart

#### use of `agg()`

###### Ex. Extract Monthly Sales and Profit

### Analysing Dataframes

- univariate analysis - boxplot, histogram, value_counts(), countplot, describe()
- bivariate analysis
    - categorial X numerical - barchart, piechart
    - 2 numerical - scatter plot
    - 2 categorial - crosstab
- multivariate - pivot table

### Univariate Analysis

#### Summary Statistics

`df.describe()` **- Generates descriptive statistics. Descriptive statistics include those that summarize the central tendency, dispersion and shape of a dataset’s distribution, excluding NaN values. Analyzes both numeric and object series, as well as DataFrame column sets of mixed data types. The output will vary depending on what is provided.**

`df.value_counts(normalize = False)` **- returns a Series containing counts of unique rows in the DataFrame**

#### Histogram

#### Box and Whisker Plot

<img src = "images/boxplot.jpg" width = 800 align = left>


### Bivariate Analysis

`pd.crosstab(index, columns, values=None, aggfunc=None normalize=False)` **- Computes a simple cross tabulation of two (or more) factors. By default computes a frequency table of the factors unless an array of values and an aggregation function are passed.**

###### Ex. Number of franchise where a product is sold across each state

###### Ex. Product and the number of time Sales Target achieved

`df.pivot_table(values=None, index=None, columns=None, aggfunc='mean')`  **- creates a spreadsheet-style pivot table as a DataFrame. The levels in the pivot table will be stored in MultiIndex objects (hierarchical indexes) on the index and columns of the result DataFrame.**

#### Barchart

###### Display Sales across products

#### Piechart

#### Scatter Plot

#### Line Chart

###### Ex. Plot line chart showing monthly sales

#### using pandas

#### using seaborn